README.md
# Cherry

Cherry is an OpenFlow controller written in Go that supports OpenFlow 1.0 and 1.3 protocols. This project is not designed for general purpose, and it instead focuses on SDN (Software-Defined Networking) for an IT service provider.

## Features

* Supports OpenFlow 1.0 and 1.3 protocols
* Focuses on compatibility with commercial OpenFlow-enabled switches
* Supports network topology that has loops in it
* Provides several northbound applications: ProxyARP, L2Switch, Floating-IP, etc.
* Provides simple plugin system for northbound applications
* RESTful API to manage the controller itself

## Supported OpenFlow Switches (Fully Tested)

* Dell Force10 S4810
* Dell Force10 S3048-ON
* Dell Force10 S3148
* Open vSwitch

## Tested OpenFlow Switches

* HP 2920G
* Pica8 P-3295
* Quanta T1048-LB9

## Requirements

* MySQL (or MariaDB) database server

## Quick Start

You can install Cherry on Docker or natively from source based on your preference. 

### Installing on Docker

* Install Docker if you don't have it on your system by following instruction: https://docs.docker.com/installation/
* Clone Cherry:

 ```$ git clone https://github.com/superkkt/cherry.git```

* Copy the template configuration: 
 
 ```$ sudo cp cherry/cherry.yaml /usr/local/etc```

* Edit the configuration file */usr/local/etc/cherry.yaml* as you want.

* Build Docker image as root:

 ```# cd cherry; docker build -t cherry .```

* Run as root:

 ```# docker run -d -p 6633:6633 -v /dev/log:/dev/log -v /your/config/cherry.yaml:/usr/local/etc/cherry.yaml cherry```

 The bind mount of /dev/log is to collect syslog messages from the container and then write to the host's syslog daemon.

* That's it! Cherry will be started in L2 switch mode.

### Installing from source

* Install Go language if you don't have it on your system by following instruction: http://golang.org/doc/install
* Clone and compile Cherry: 

 ```$ go get github.com/superkkt/cherry```

* Copy the compiled binary and template configuration: 
 
 ```$ sudo cp $GOPATH/bin/cherry /usr/local/bin```
 
 ```$ sudo cp $GOPATH/src/github.com/superkkt/cherry/cherry.yaml /usr/local/etc```

* Edit the configuration file */usr/local/etc/cherry.yaml* as you want.

* Run:

 ```$ /usr/local/bin/cherry &```

* That's it! Cherry will be started in L2 switch mode.

## Copyright and License

```
Copyright (C) 2015 Samjung Data Service, Inc. All rights reserved.
Kitae Kim <superkkt@sds.co.kr>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
```


Changelog
* v0.14.2 (2019-09-20)

3faead4 Fix the unexpected master change (See #44)

* v0.14.1 (2019-09-03)

b321987 Remove password plain text log (See #43)

* v0.14.0 (2019-08-02)

fad9247 Change ui server middleware order (See #42)
e02444b Change name of host table timestamp column (See #41)
7ee607a Revert "Set hard-timeout to clear stale flows"
71199c6 Change data column type in log table (See #38)
3b8fe3e Apply LDAP (See #31)
715f48f Change validation of input length (See #36)
6a81a12 Change timestamp column option in host table (See #37)
9cf45b8 Apply TOTP (See #30)
6f28b01 Prevent gateway from being set as reserved IP (See #35)
4faeba4 Change host update API (See #26)
efaab76 Change IP structure (See #27)
8f0eb70 Add get host API (See #34)
040d871 Add gateway within network (See #25)
501c3d3 Add network search (See #24)
8740e05 Add spec feature within host (See #29)
2e8ed92 Add Component UI API (See #28)
3f301a2 Add Category UI API (See #23)
80606bb Add Log UI API (See #18)
519ecce Fix a bug to add an invalid IP address to the database when adding a network (See #33)
c4c6fdf NULL MAC address to disconnect a host from the network (See #32)
3504fd6 Dedicated IP address for ARP host discovery (See #12)
cb92e4a Use reserved IP addresses (TEST-NET-1) for DHCP and ARP discovery
79d95da Validate TCP and UDP packet lengths
a5415ba Fix the broken UDP checksum
6ce8b01 Implement a DHCP server
c01b487 refactoring: apply the custom response writer to the UI API
cf82a97 revise
abf5e1a refactoring: apply pagination with structure
fbf94f0 refactoring: ui transaction returns instance reference
936f21d Add Host List API (See #20)
74ff82e fix add, update host
9e013c9 refactoring: call database logics in a closure
b9b85d4 refactoring: apply the custom response writer to the core API
726f009 refactoring: custom response writer with logging
58362ed Add Host UI API (See #20)
c31fdaa Add VIP UI API (See #19)
a13a7dc Refactoring: re-organize the source tree structure of the API package
5667f06 Add Switch UI API (See #14)
0090f4b Add Network UI API (See #15)
1729ffa Add Host group UI API (See #17)
4e4188f Add User UI API (See #16)
b10bfa7 Add extra comment for time interval of the flow manager
a4aaac9 Set hard-timeout to clear stale flows
0c939a7 ARP announcer (See #22)
cd32572 Devide UI API server (See #13)
9af1b5b Fix the broken test case
e2e8add Replace godep with dep to vendor libraries
9cddf5f Add extra debug logs for broadcast packets
b1e6584 Update the buffered socket stream to avoid a potential internal buffer corruption
ef776e5 Add the GitLab CI
f4f235f Fix a typo
1fa1d05 Fix broken test cases
146b172 Fix the IP conflict by ARP probes (See #12)

* v0.13.5 (2018-05-15)

06fe675 Discovery: workaround for old OSes that have a broken TCP/IP stack (See #6)

* v0.13.4 (2018-04-22)

4767d4c Fix the node discovery bug (See #5)

* v0.13.3 (2018-04-22)

619eefc Revert "Decrease the PACKET_IN read buffer size"

* v0.13.2 (2018-04-20)

05f3fc0 Adjust log levels
412988e Use string parameters in the YAML config file
1114081 MySQL: validate the cluster address
fe7f912 Update the MySQL driver
c3b6288 mysql: new cluster dialer and query function that pass Tx instead of DB
a760e5e Ignore the WRITE fsnotify operation to avoid reading empty config
d4d48f7 Increase the flow idle timeout
df3e5ed Make a switch drops all incoming packets (except ARP and LLDP) for a while on start up
0bfdb5b Sort networks by address
3809a8c Re-ordering DB tables in the MySQL schema
9eb8066 Decrease the ARP probe interval
480d6b5 Fix missing time interval change
7cfdccb Decrease the flow manager interval
6f8780c Add extra comments
151ab37 Decrease the device explorer interval to 1 minute
0e11b0d Update a debug message
1fc910c Decrease ARP sender's loop interval to 1.5 sec.
5472f62 Add a comment for time buffering
750627a Drop the some PACKET_INs for proper device setup on start up
0670200 Drop the packets before ethernet parsing if the device is not yet ready
e7d0e91 Decrease ARP sender's loop interval
3f3fbc3 Send ARP probes immediately on start up for fast discovery
a77d304 Fix aaa0e4fd797 as OF1.0 does not support the cookie mask
5ade735 Deny the API request if we are not the master controller
5e2633e Keep the ARP flow unless the switch device reconnects
aaa0e4f Flow cache to avoid duplicate and repeated flow installation (See #1)
f8bb64f Minor refactoring
109e393 Refactoring of the switch device negotiation procedure
c527c28 Add extra debug messages
e8d4b50 Decrease the timeout for a staled host
6d42974 Decrease the PACKET_IN read buffer size
78f098a Use FLOW_ADD instead of FLOW_MODIFY for a broken switch device (i.e., Dell S3148) (See #1)
7770802 Log DPID when we get an openflow error packet
018b460 Ignore the invalid echo responses

* v0.13.1 (Sat Aug 05 16:45:00 2017 +0900)

075e31c Ignore the ARP reply received from an edge among switches

* v0.13.0 (Fri Aug 04 23:37:48 2017 +0900)

fe412ce Remove duplicated link down check
5bfbf68 Decrease the flow manager interval to 1 minute
0130939 Flow manager to keep the flows update
b92618c Fix the deadlock between the topology and the device
698ce6a Minor refactoring
7a88c77 Set idle and hard flow timeouts from the config
ae4e791 Log flow removed timestamp
13ba29f Remove all the flow histories on the device down event
fd8b1a4 Warning log on the port down event
72b66c2 Backoff-delayed ARP announcer for all the registered hosts
fbc8759 Backoff-delayed ARP announcer
0509b0c Invalidate the ARP caches when a host is removed
9d0500e Fix the broken VIP toggling by repeated toggles in a moment
d468086 Minor refactoring of README
f7cee84 Update README for the new configuration file
ae3ee97 Fix the broken VIP toggling by the Discovery module
ad5b2d9 Revert a weird patch 601ac8055
ce25c01 Minor refactoring by function renaming
daaeaed Log switch device up/down events
a34bdce Fix the broken Dockerfile
ee8f7a4 Flow history on the database
e46bce5 Adjust switch port numbers based on its first printed port number

main.go
/*
 * Cherry - An OpenFlow Controller
 *
 * Copyright (C) 2015 Samjung Data Service, Inc. All rights reserved.
 * Kitae Kim <superkkt@sds.co.kr>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package main

import (
	"context"
	"flag"
	"fmt"
	"net"
	"os"
	"os/signal"
	"runtime"
	"strings"
	"syscall"
	"time"

	"github.com/superkkt/cherry/database"
	"github.com/superkkt/cherry/election"
	"github.com/superkkt/cherry/network"
	"github.com/superkkt/cherry/northbound"

	"github.com/fsnotify/fsnotify"
	"github.com/pkg/errors"
	"github.com/superkkt/go-logging"
	"github.com/superkkt/viper"
)

const (
	programName     = "cherry"
	programVersion  = "0.13.5"
	defaultLogLevel = logging.INFO
)

var (
	logger            = logging.MustGetLogger("main")
	loggerLeveled     logging.LeveledBackend
	showVersion       = flag.Bool("version", false, "Show program version and exit")
	defaultConfigFile = flag.String("config", fmt.Sprintf("/usr/local/etc/%v.yaml", programName), "absolute path of the configuration file")
)

func main() {
	runtime.GOMAXPROCS(runtime.NumCPU())
	flag.Parse()
	if *showVersion {
		fmt.Printf("Version: %v\n", programVersion)
		os.Exit(0)
	}

	initConfig()
	if err := initLog(getLogLevel(viper.GetString("default.log_level"))); err != nil {
		logger.Fatalf("failed to init log: %v", err)
	}

	ctx, cancel := context.WithCancel(context.Background())
	db, err := database.NewMySQL()
	if err != nil {
		logger.Fatalf("failed to init MySQL database: %v", err)
	}

	observer := initElectionObserver(ctx, db)

	controller := network.NewController(db, observer)
	manager, err := createAppManager(db)
	if err != nil {
		logger.Fatalf("failed to create application manager: %v", err)
	}
	manager.AddEventSender(controller)

	initSignalHandler(controller, manager, cancel)

	listen(ctx, viper.GetInt("default.port"), controller, observer)
}

func initConfig() {
	viper.SetConfigFile(*defaultConfigFile)
	// Read the config file.
	if err := viper.ReadInConfig(); err != nil {
		logger.Fatalf("failed to read the config file: %v", err)
	}
	// Watching and re-reading config file whenever it changes.
	viper.OnConfigChange(func(e fsnotify.Event) {
		// Ignore the WRITE operation to avoid reading empty config.
		if e.Op != fsnotify.Write {
			return
		}

		if loggerLeveled != nil {
			// Set log level for all modules
			loggerLeveled.SetLevel(getLogLevel(viper.GetString("default.log_level")), "")
		}
	})
	viper.WatchConfig()
	if err := validateConfig(); err != nil {
		logger.Fatalf("failed to validate the configuration: %v", err)
	}
}

func validateConfig() error {
	if port := viper.GetInt("default.port"); port <= 0 || port > 0xFFFF {
		return errors.New("invalid default.port")
	}
	if len(viper.GetString("default.log_level")) == 0 {
		return errors.New("invalid default.log_level")
	}
	if len(viper.GetString("default.applications")) == 0 {
		return errors.New("invalid default.applications")
	}
	if len(viper.GetString("default.admin_email")) == 0 {
		return errors.New("invalid default.admin_email")
	}
	vlanID := viper.GetInt("default.vlan_id")
	if vlanID < 0 || vlanID > 4095 {
		return errors.New("invalid default.vlan_id in the config file")
	}

	return nil
}

func initElectionObserver(ctx context.Context, db *database.MySQL) *election.Observer {
	observer := election.New(db)
	go func() {
		if err := observer.Run(ctx); err != nil {
			logger.Fatalf("failed to run the election observer: %v", err)
		}
		logger.Debugf("election observer terminated")
	}()

	return observer
}

func initSignalHandler(controller *network.Controller, manager *northbound.Manager, cancel context.CancelFunc) {
	go func() {
		c := make(chan os.Signal, 5)
		// All incoming signals will be transferred to the channel
		signal.Notify(c)

		// Infinte loop.
		for {
			s := <-c
			if s == syscall.SIGTERM || s == syscall.SIGINT {
				// Graceful shutdown
				logger.Warning("Shutting down...")
				cancel()
				// Timeout for cancelation
				time.Sleep(5 * time.Second)
				os.Exit(0)
			} else if s == syscall.SIGHUP {
				fmt.Println("* Controller status:")
				fmt.Println(controller.String())
				fmt.Printf("\n* Manager status:\n")
				fmt.Println(manager.String())
			}
		}
	}()
}

func initLog(level logging.Level) error {
	backend, err := newSyslog(programName)
	if err != nil {
		return err
	}
	backend = logging.NewBackendFormatter(backend, logging.MustStringFormatter(`%{level}: %{shortpkg}.%{shortfunc}: %{message}`))

	loggerLeveled = logging.AddModuleLevel(backend)
	// Set log level for all modules
	loggerLeveled.SetLevel(level, "")
	logging.SetBackend(loggerLeveled)

	return nil
}

func getLogLevel(level string) logging.Level {
	level = strings.ToUpper(level)
	ret, err := logging.LogLevel(level)
	if err != nil {
		logger.Infof("invalid log level=%v, defaulting to %v..", level, defaultLogLevel)
		return defaultLogLevel
	}

	return ret
}

func listen(ctx context.Context, port int, controller *network.Controller, observer *election.Observer) {
	type KeepAliver interface {
		SetKeepAlive(keepalive bool) error
		SetKeepAlivePeriod(d time.Duration) error
	}

	listener, err := net.Listen("tcp", fmt.Sprintf(":%v", port))
	if err != nil {
		logger.Errorf("failed to listen on %v port: %v", port, err)
		return
	}
	defer listener.Close()

	// Connection dispatcher.
	f := func(c chan<- net.Conn) {
		for {
			conn, err := listener.Accept()
			if err != nil {
				logger.Errorf("failed to accept a new connection: %v", err)
				continue
			}
			logger.Infof("new device is connected from %v", conn.RemoteAddr())

			// Only the master controller can serve the connections!
			if observer.IsMaster() == false {
				logger.Warningf("disconnecting the newly connected device (%v) because we are not the master controller!", conn.RemoteAddr())
				conn.Close()
				continue
			}

			// Pass the new connection into the backlog queue.
			c <- conn
		}
	}
	backlog := make(chan net.Conn, 32)
	go f(backlog)

	// Infinite loop
	for {
		select {
		case <-ctx.Done():
			logger.Debug("terminating the main listener loop...")
			return
		case conn := <-backlog:
			logger.Debug("fetching a new connection from the backlog..")
			if v, ok := conn.(KeepAliver); ok {
				logger.Debug("trying to enable socket keepalive..")
				if err := v.SetKeepAlive(true); err == nil {
					logger.Debug("setting socket keepalive period...")
					// Makes a broken connection will be disconnected within 45 seconds.
					// http://felixge.de/2014/08/26/tcp-keepalive-with-golang.html
					v.SetKeepAlivePeriod(time.Duration(5) * time.Second)
				} else {
					logger.Errorf("failed to enable socket keepalive: %v", err)
				}
			}
			controller.AddConnection(ctx, conn)
		}
	}
}

func createAppManager(db *database.MySQL) (*northbound.Manager, error) {
	manager, err := northbound.NewManager(db)
	if err != nil {
		return nil, err
	}

	apps, err := parseApplications()
	if err != nil {
		return nil, errors.Wrap(err, "failed to parse applications")
	}
	for _, v := range apps {
		if err := manager.Enable(v); err != nil {
			return nil, errors.Wrap(err, fmt.Sprintf("enabling %v", v))
		}
	}

	return manager, nil
}

func parseApplications() ([]string, error) {
	// Remove spaces, and then split it using comma
	tokens := strings.Split(strings.Replace(viper.GetString("default.applications"), " ", "", -1), ",")
	if len(tokens) == 0 {
		return nil, errors.New("empty application")
	}

	return tokens, nil
}
